### PR TITLE
Add Kriegspiel

### PR DESCRIPTION
**Pull request by iamcxds**
#### Checklist

* [x] Use a separate branch in your local repo (not `master`).
* [x] I've read and agree with the [contribution guidelines](https://www.freeboardgames.org/docs/?path=/story/documentation-contributing--page).

### CODE REVIEW

File: web/src/games/kriegspiel/ai.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    return cell;
+  }
+  makeMove(playerID: string, cell: number) {
+    return { action: { type: 'MAKE_MOVE', payload: { type: 'clickCell', args: [cell], playerID } } };
+  }
+  randomNumber(min: number, max: number) {
+    // return Math.floor(Math.random() * (max - min + 1) + min);  // https://github.com/babel/minify/issues/904
+    return (Math.random() * (max - min + 1) + min) << 0;
+  }
+} */
```
Comment (vdfdev) [inline web/src/games/kriegspiel/ai.ts:39]: Please remove dead code from comments.

File: web/src/games/kriegspiel/config.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { IGameConfig } from 'gamesShared/definitions/game';
+import { Kriegspiel } from './Game';
+import { Board } from './Board';
+
+const config: IGameConfig = {
+  bgioGame: Kriegspiel,
+  bgioBoard: Board,
+  debug: true,
```
Comment (vdfdev) [inline web/src/games/kriegspiel/config.ts:8]: We probably want to leave debug: false before merging.

File: web/src/games/kriegspiel/index.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const Thumbnail = require('./media/thumbnail.png?lqip-colors');
+import { GameMode } from 'gamesShared/definitions/mode';
+import { IGameDef, IGameStatus } from 'gamesShared/definitions/game';
+import instructions from './instructions.md';
+
+export const kriegspielGameDef: IGameDef = {
+  code: 'kriegspiel',
+  name: "Guy Debord's Kriegspiel",
```
Comment (vdfdev) [inline web/src/games/kriegspiel/index.ts:8]: We generally dont allow using the original game name for copyright reasons. However, this game doesn't seem to be historic and not commercialized anymore, so we might be OK leaving the game name as is, and if someone complains we can change it later. https://en.wikipedia.org/wiki/A_Game_of_War

Comment (vdfdev) [reply web/src/games/kriegspiel/index.ts:8]: (no action needed)

File: web/src/games/kriegspiel/instructions.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# Le Jeu de la Guerre
+
+Try it on the [GitHub page](https://iamcxds.github.io/kriegspiel).
+
+A WEB implementation of Guy Debord's A Game of War ([Le Jeu de la Guerre](https://www.wikiwand.com/fr/Le_Jeu_de_la_guerre_(livre))). Powered by [boardgame.io](https://boardgame.io/). The graphic is based on svg+emoji, with [Pico8's Palette](https://pico-8.fandom.com/wiki/Palette).
```
Comment (vdfdev) [inline web/src/games/kriegspiel/instructions.md:5]: This seems a bit out of place for users going to FBG? They dont really care we use boardgame.io and etc, I think you can delete this first section...

File: web/src/games/kriegspiel/instructions.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  - 😅 Out of Supply
+  - 💥 Battle
+
+### Rules
+
+Please check the [rules](http://r-s-g.org/kriegspiel/rules.php) here.
+
+## The Other Implementation
+
+There is another digital implementation: [Kriegspiel](http://r-s-g.org/kriegspiel/index.php), made by Alex Galloway. There are an [interview](https://www.youtube.com/watch?v=CGjt8po_y4I) and a [playthrough](https://www.youtube.com/watch?v=4l2M6vpWLAw) by Fred Serval. Unfortunately, it's only available on Mac and IOS, which inspires me to make a WEB version.
```
Comment (vdfdev) [inline web/src/games/kriegspiel/instructions.md:48]: I think the "try it on github page" would be better here.

File: web/src/games/kriegspiel/Board.tsx [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+  getSuppliedCells,
+  exportGame,
+} from './Game';
+
+//import { useGesture } from '@use-gesture/react'
+
+/* const getWinner = (ctx: Ctx) => {
+  if (!ctx.gameover) return null;
+  else return spanBGColor(<>The winner is Player {ctx.gameover.winner}!</>, fictionColor(ctx.gameover.winner));
+}; */
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Board.tsx:29]: Please remove dead code.

File: web/src/games/kriegspiel/Board.tsx [hunk 7]
```diff
@@ -0,0 +1,770 @@
...
+        setMapScale(newScale);
+      },
+
+    },
+    {
+      target: boardRef,
+      eventOptions: { passive: false },
+      preventDefault: true,
+    }
+  ); */
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Board.tsx:182]: Please remove dead code.

File: web/src/games/kriegspiel/Board.tsx [hunk 8]
```diff
@@ -0,0 +1,770 @@
...
+
+  //render Main board
+
+  const gameBoard = (
+    <svg
+      viewBox={`-0.6 -0.6 ${BoardSize.mx + 1.2} ${BoardSize.my + 1.2}`}
+      ref={boardRef}
+      //onWheel={myOnWheel} onMouseMove={drag} onMouseDown={startDarg} onMouseUp={endDarg} onMouseLeave={endDarg}
+      //onTouchStart={startDarg} onTouchEnd={endDarg} onTouchCancel={endDarg}
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Board.tsx:191]: Please remove dead code.

File: web/src/games/kriegspiel/Board.tsx [hunk 9]
```diff
@@ -0,0 +1,770 @@
...
+              lines.length > 1 &&
+              gTranslate(
+                drawLine(lineLst[0], lineLst[lineLst.length - 1], fictionColor('0'), 0.05, '0.5, 0.1'),
+                -0.05,
+                -0.05,
+              )
+            );
+          }),
+        )}
+        {getDirSuppliedLines(G, '1')[1].map((lines) =>
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Board.tsx:254]: This function is too large and too hard to read. Please break it down into smaller React components. Each .map() function here is a good candidate for a separate component.

File: web/src/games/kriegspiel/Board.tsx [hunk 10]
```diff
@@ -0,0 +1,770 @@
...
+  white: '#fff1e8',
+  red: '#ff004d',
+  orange: '#ffa300',
+  yellow: '#ffec27',
+  green: '#00e436',
+  blue: '#29adff',
+  lavender: '#83769c',
+  pink: '#ff77a8',
+  light_peach: '#ffccaa',
+};
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Board.tsx:770]: If possible, it is more readable to use CSS instead, using .module.css, here is an example: https://github.com/freeboardgames/FreeBoardGames.org/blob/b596fda05f77b5ac208e842769fcd25f6d40f665/web/src/games/zooparade/components/bdeck.tsx#L13

File: web/src/games/kriegspiel/instructions.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
+# Le Jeu de la Guerre
```
Comment (vdfdev) [inline web/src/games/kriegspiel/instructions.md:1]: This game doesnt look like it is going to play well in mobile phones. If you don't plan on supporting mobile phones, please highlight it here on the instructions.

File: web/src/games/kriegspiel/Game.ts [hunk 12]
```diff
@@ -0,0 +1,849 @@
...
+  readonly objType: ObjType; // functional type
+  readonly objRender: string; //emoji
+  readonly speed: number;
+  readonly range: number;
+  readonly offense: number;
+  //readonly offenseOnCharge?:number, +3 for "Cavalry"
+  readonly defense: number;
+  readonly canAddDef: boolean;
+  //readonly defenseInMountain:number, +2 for "Infantry" and "Artillery"
+  //readonly defenseInFort:number +4
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Game.ts:733]: Please remove unused code.

File: web/src/games/kriegspiel/Game.test.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  p0.moves.movePiece(77, 76);
+
+  p0.events.endTurn();
+
+  // Player 1's state reflects the moves made by Player 0.
+  const gameData =
+    '|32|🏰|6|🎪.0|19|⛰️|⛰️|⛰️|⛰️|13|🚩.0|5|🎪.0|1|⛰️|19|🏇.0|2|🎉.0|1|⛰️|23|💂.0|🛣️|17|🏇.0|🏇.0|🚚.0|💂.0|💂.0|1|💂.0|⛰️|19|💂.0|🏇.0|💂.0|💂.0|💂.0|⛰️|10|🏰|9|💂.0|3|⛰️|2|🏰|3|🏇.1|14|🚀.0|7|💂.1|💂.1|💂.1|21|💂.1|3|🎉.1|21|🏰|💂.1|💂.1|🏇.1|🏇.1|8|🏰|11|💂.1|💂.1|💂.1|🏇.1|17|⛰️|⛰️|⛰️|⛰️|⛰️|⛰️|🚚.1|23|🚀.1/🛣️.1|6|🚩.1/🏰|17|⛰️|24|⛰️|24|⛰️|36|🎪.1|19|🎪.1|';
+  expect(p1.getState().G.cells).toEqual(Game.loadPieces(gameData));
+  expect(p0.getState().G.cells).toEqual(Game.loadPieces(gameData));
+});
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Game.test.ts:58]: Ideally, you would be able to test more corner cases using unit tests, to make sure there is no bug in the rules.

**Review by vdfdev — CHANGES_REQUESTED** [review]

File: web/src/games/kriegspiel/Board.tsx [hunk 14]
```diff
@@ -0,0 +1,770 @@
...
+            navigator.clipboard.writeText(gameData);
+          }}
+        />
+        <input type="button" value="Remove Data" onClick={() => setGameData('')} />
+      </form>
+    </div>
+  );
+  // render all
+  return (
+    <main>
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Board.tsx:641]: Please make sure you wrap your board with <GameLayout, like here:
https://github.com/freeboardgames/FreeBoardGames.org/blob/4361c9f25e431b91ce9ff99cba78855674b00046/web/src/games/tictactoe/board.tsx#L93

and that you pass gameOver when the game is over. This will allow users to navigate back to the main menu and replay your game (or switch to another game) when it is over.

File: web/src/games/kriegspiel/Board.tsx (hunk 7 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/Board.tsx:182]: this code is for mobile phone UI control

File: web/src/games/kriegspiel/instructions.md (hunk 11 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/instructions.md:1]: I have mobile phone UI control support, by using the react-use-gesture, should I load this library?

File: web/src/games/kriegspiel/instructions.md (hunk 4 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/instructions.md:5]: And by the way, how to insert images in instructions.md?

File: web/src/games/kriegspiel/Board.tsx (hunk 7 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/Board.tsx:182]: how to add [this libarary](https://github.com/pmndrs/use-gesture)? The simply installing doesn't work

File: web/src/games/kriegspiel/Board.tsx (hunk 10 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/Board.tsx:771]: I find it is easier to control by script in this way.

File: web/src/games/kriegspiel/instructions.md (hunk 11 above)
Comment (vdfdev) [reply web/src/games/kriegspiel/instructions.md:1]: Ok, lets follow up on this later after we upgrade react.

File: web/src/games/kriegspiel/ai.ts (hunk 1 above)
Comment (vdfdev) [reply web/src/games/kriegspiel/ai.ts]: (not resolved)

There is no reason to have commented out code here. If you are leaving this just to use it later, you can look at the git history.

File: web/src/games/kriegspiel/config.ts (hunk 2 above)
Comment (vdfdev) [reply web/src/games/kriegspiel/config.ts]: (not resolved)

File: web/src/games/kriegspiel/Game.test.ts (hunk 13 above)
Comment (vdfdev) [reply web/src/games/kriegspiel/Game.test.ts:58]: Random AI won't automatically run in every commit and every dependency update on this repository, so it won't catch if the game breaks in the future by dependency upgrades, like boardgame.io. If you want your game to not be broken in the future, you probably want to add more tests.

File: web/src/games/kriegspiel/Game.ts [hunk 15]
```diff
@@ -0,0 +1,828 @@
...
+  const obj = G.cells[stCId];
+  //check obj is on stCells,  in move range
+  return obj !== null && moveRange(G, stCId, obj.speed).includes(edCId);
+}
+function moveRange(G: GameState, stCId: CellID, speed: number = 1): CellID[] {
+  let result = [stCId];
+  for (let i = 0; i < speed; i++) {
+    //for each steps target cell is empty and is not mountain,
+    // first filter out further cells
+    // let result be 1 block larger
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Game.ts:469]: Please remove dead code in comments.

File: web/src/games/kriegspiel/Game.ts [hunk 16]
```diff
@@ -0,0 +1,828 @@
...
+      //filter out retreating units in offense
+      !(isOffense && (obj.retreating || obj.offense === 0))
+    );
+  });
+  /* filterCId(G.cells,(obj,id)=>
+  //obj is in range, supplied, belongs to the chosen player,
+  (isInRange(pos,CId2Pos(id),obj)&&obj.supplied&&obj.belong===player)
+  //filter out retreating units in offense
+  &&!(isOffense&&obj.retreating)
+  ) */
```
Comment (vdfdev) [inline web/src/games/kriegspiel/Game.ts:599]: Please remove dead code in comments.

**Review by vdfdev — CHANGES_REQUESTED** [review]

File: package.json [hunk 17]
```diff
@@ -27,6 +27,7 @@
     "web:dev": "cd web/ && yarn run dev:internal"
   },
   "dependencies": {
+    "@use-gesture/react": "^10.2.20",
```
Comment (vdfdev) [inline package.json:30]: Remove this. This is in the wrong package. Any dependency should be added to web/package.json, not the root package.json

**Review by vdfdev — CHANGES_REQUESTED** [review]

File: web/src/games/kriegspiel/ai.ts (hunk 1 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/ai.ts]: I already deleted it if you check the latest update

File: web/src/games/kriegspiel/config.ts (hunk 2 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/config.ts]: please check the latest version

File: package.json (hunk 17 above)
Comment (iamcxds) [reply package.json:30]: Sorry for making this error, now this library is installed and functions properly.

File: web/src/games/kriegspiel/Game.test.ts (hunk 13 above)
Comment (iamcxds) [reply web/src/games/kriegspiel/Game.test.ts:58]: Ok, I add a test for attack move, and I think that is enough. There are only two moves for the game: movePiece and attack.

**Review by vdfdev — APPROVED** [review]
